Differentiate between macOS and iOS minimal supported version for CMake builds

discnl opened the pull request.

discnl commented:
This is the architecture mismatch Maarten referred to in https://github.com/wxWidgets/wxWidgets/commit/e4437c49749bf2a30b9f639a85061eea99df6aa7#r45099359 :

```
ld: warning: ignoring file /usr/local/lib/libwx_osx_iphoneu_core-3.1.dylib, building for iOS-arm64 but attempting to link with file built for iOS-armv7

ld: warning: ignoring file /usr/local/lib/libwx_baseu-3.1.dylib, building for iOS-arm64 but attempting to link with file built for iOS-armv7

```

I finally managed to reproduce it locally as well and will look into it further.

vadz commented on CMakeLists.txt:
```diff
@@ -26,10 +26,16 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
     set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release")
 endif()
 
+# This block, particularly the versions used, should be kept in sync with
+# samples/minimal/CMakeLists.txt.
```
Could we have some `apple_target.cmake` and include it from both places?

discnl replied:
Likely, but earlier I decided not to look into that for dependency reasons mentioned in the commit message. For that same reason I didn't, from a [user's POV](https://github.com/wxWidgets/wxWidgets/blob/c069a62bd6e2a6fa6edc84a12b5fb466cd3b0386/samples/minimal/CMakeLists.txt#L11), pollute minimal's CMakeLists.txt by repeating the "keep in sync" message there as well.

The sample includes the wxWidgets use file but that file lives outside of wx and any changes would always require the latest CMake so that doesn't seem like an option.

There's an `OPTIONAL` keyword for includes, this works from samples/minimal/CMakeLists.txt:

`include(apple_target.cmake OPTIONAL)`

I think that's the least intrusive (when not wanting to keep two pieces of code in sync) but perhaps it's better to just have the file under build/cmake/ , despite a user seeing that path then.

Of course it's also possible to force inclusion as usual and risk errors (which is probably annoying if one is targeting non-Apple), it just wouldn't be my preference as not setting the target OS shouldn't by default result in a problem for macOS and iOS (right now it is for the latter due to arch differences but that's another issue).

vadz replied:
OK, I didn't think about it, but you're right, this is not ideal for the users, so let's keep things simple and live with this small amount of duplicated code. Sorry for not thinking it through before making this comment.

discnl replied:
No problem at all and thanks for reviewing despite, like me, not knowing about CMake (my knowledge doesn't go much further then pressing the... Generate button once again in case of errors).
No changes are needed (even if minimal iOS may have to change it can be done later on), but I'll wait a few more days to see if Maarten has something to say just in case.

vadz commented:
Thanks! This looks correct to me but, of course, I don't know anything about this stuff, so I'd appreciate a green light from @MaartenBent before merging it.

discnl commented:
> Thanks! This looks correct to me but, of course, I don't know anything about this stuff, so I'd appreciate a green light from @MaartenBent before merging it.

After required changes I don't mind merging (or rebasing, for me it depends per situation) this myself. I should probably start using assignee to indicate that and reduce any duplicate effort.

MaartenBent reviewed (approved):
Looks fine to me.

There is another, different, check for iOS/iPhone in `toolkit.cmake` using `CMAKE_OSX_SYSROOT`. Maybe we can use `CMAKE_SYSTEM_NAME` here as well?

discnl commented:
Thanks for reviewing. I think it first depends on if it's checking for merely iOS compilation there or actually checking if an iOS device is targeted?
When using a simulator the SDK path doesn't contain "iphoneos" (but rather "iphonesimulator") so if it's trying to check for iOS compilation it looks like it would be wrong. I couldn't find how/if cross-compilation is supported with CMake for wx and had no luck trying a quick configure-based CMake build for iOS so can't check it further for now. At least from looking at `IPHONE` hits most are related to platform checks but not all.

discnl commented on CMakeLists.txt:
```diff
@@ -28,8 +28,12 @@ endif()
+    if(CMAKE_SYSTEM_NAME STREQUAL "iOS")
+        set(CMAKE_OSX_DEPLOYMENT_TARGET 10.0 CACHE STRING "macOS Deployment Target")
+    else()
+        set(CMAKE_OSX_DEPLOYMENT_TARGET 10.10 CACHE STRING "macOS Deployment Target")
```
Seeing this [change]( https://github.com/wxWidgets/wxWidgets/commit/4fb1764d6b1825af8e089b151650ceaa76e79666#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL32-R32) and also now knowing that the literal string for `set` is a docstring: Could the (second) instance of `macOS` be replaced with `iOS` here?

MaartenBent replied:
Sure, it is just a description of the variable.

discnl committed:
CMake: Separate minimal support for iOS and macOS

Differentiate between the minimal supported version of macOS (10.10)
and iOS (12.0) instead of setting CMAKE_OSX_DEPLOYMENT_TARGET to
(macOS-only) 10.10 for both platforms.

discnl committed:
CMake: Fix out of sync iOS/macOS deployment target

The minimal sample doesn't have an implicit deployment target set, while
the library does through CMakeLists.txt in root. This can result in iOS
link errors when not explicitly setting CMAKE_OSX_DEPLOYMENT_TARGET as
the sample will then use a target OS with an architecture that may not
be present in the library with its current target OS version.

Fix by copying the part setting CMAKE_OSX_DEPLOYMENT_TARGET in the
library to the minimal sample as well. Both parts should be kept in sync
but that still seemed preferable to additionally polluting an example
file for CMake usage with a dependency by introducing a common include
file.

discnl committed:
Travis CI: Use default target for CMake iOS build

Remove an explicitly set CMAKE_OSX_DEPLOYMENT_TARGET from the CMake iOS
build in order to use the default value which is now also set to an iOS
instead of macOS version.
